crates/core/src/util.rs

```diff
@@ -256,29 +255,33 @@ impl TimeSource for SystemTime {
     }
 }
 
+use std::cell::RefCell;
+use std::rc::Rc;
+
 #[cfg(test)]
 #[derive(Clone)]
 pub struct MockTimeSource {
-    current_instant: Instant,
+    current_instant: Rc<RefCell<Instant>>,
 }
 
 #[cfg(test)]
 impl MockTimeSource {
     pub fn new(start_instant: Instant) -> Self {
         MockTimeSource {
-            current_instant: start_instant,
+            current_instant: Rc::new(RefCell::new(start_instant)),
         }
     }
 
-    pub fn advance_time(&mut self, duration: time::Duration) {
-        self.current_instant += duration;
+    pub fn advance_time(&self, duration: Duration) {
+        let mut instant = self.current_instant.borrow_mut();
+        *instant += duration;
     }
 }
 
 #[cfg(test)]
 impl TimeSource for MockTimeSource {
     fn now(&self) -> Instant {
```

Comment:
is fine to take  &mut here, we are not going to use under a scenario where it shouldn't be possible